### ISSUE TITLE
Add option to disable the automated settings of git env vars or make it possible to customise them

### ISSUE THREAD

**Issue by ntimo** [coder/coder#2665]
# Current behavior
The coder agent sets the following env vars in the environment of the user.
- GIT_COMMITTER_EMAIL
- GIT_COMMITTER_NAME
- GIT_AUTHOR_EMAIL

# Why can this be problematic?
It could be possible that users have a different email for their coder account that they don't want to leak in git commits or a different user name

# Suggested solution
Make it possible to:
- Option A: Disable the setting of these env vars all together
- Option B: Make it possible to customize them

**Comment by bpmct** [comment]
Option C: Use our git/ssh middleware to detect if `user.email` is not set and offer to write a default gitconfig with a `yes/no` prompt. The environment variables also take precedent over a gitconfig, so the comitter email essentially "hardcoded".

**Comment by ketang** [comment]
#2551 is related

**Comment by ketang** [comment]
My workaround was to override these environment variables in a dot file.  Not exactly elegant, but it gets the job done.

**Comment by dwahler** [comment]
Yeah, I think most people will expect to be able to configure this in `.gitconfig`, either manually or via a dotfiles repo. I'm not sure it's common knowledge that the environment variables silently take precedence (I certainly didn't realize it) so I would prefer to avoid setting them at all.

I like [Option C](https://github.com/coder/coder/issues/2665#issuecomment-1166633603) above, except that I'm not sure about interrupting the normal flow with a yes/no prompt. IMO it's probably enough to just print a clear message when automatically populating `.gitconfig`, saying what changes we're making.

It would also be nice to have a per-user setting to disable this behavior and/or control which name/address are used. That probably depends on having a general-purpose way to store per-user configuration settings.

**Comment by f0ssel** [comment]
I'm not convinced we need to do anything here, we provide users to do this however they want via dot files and any other template customizations, so my vote is to just remove this behavior entirely.

**Comment by dwahler** [comment]
That certainly has the advantage of being easy: #2980 

If we go for that option as an immediate fix, I would still propose keeping this issue open, because automatically setting up Git is convenient and it would be nice to support it in a more configurable way in the future.

**Comment by ketang** [comment]
I don't think taking this away is the best choice.  Seamless and automatic git setup is nice.  Even with us getting it wrong I still prefer it.

**Comment by dwahler** [comment]
Yeah, I would prefer to avoid getting rid of it myself.

#2981 is an alternate approach that configures these values in `~/.gitconfig`. It addresses some of the issues discussed above but still has a few limitations:

* Can't control which name and email address are automatically configured (although you can change them afterward).
* There's no message informing user of the config change, since we don't currently have a good way to directly expose such messages from the agent to the user. (We could write it to a log file, but I don't think that's any more attention-grabbing than the config change itself.)

Also, I noticed in passing that our current dotfiles implementation currently ignores all files in the dotfiles repo whose names start with `.git`, including `.gitconfig`. If we decide to support configuring `.gitconfig` through dotfiles, then I don't think we need to do anything else for the dotfiles version to correctly take precedence.

**Comment by bpmct** [comment]
#2981 seems reasonable to me as it will only set a gitconfig if one is not present and it's possible to edit with other values. That is much better than a hardcoded default.

> Also, I noticed in passing that our current dotfiles implementation currently ignores all files in the dotfiles repo whose names start with .git, including .gitconfig. If we decide to support configuring .gitconfig through dotfiles, then I don't think we need to do anything else for the dotfiles version to correctly take precedence.

@f0ssel what do you think about removing this behavior from `coder dotfiles`? It seems to be quite normal to set git preferences in dotfiles: https://sourcegraph.com/search?q=context:global+repo:dotfiles+file:%5C.git%24&patternType=regexp
